CSS/lazy/hmr: asset resource referenced in CSS isn't available to lazy loaded javascript chunk

### Have you used AI?

None

### Bug Description

When the same asset is included in both .css file (with `experiments.css`) and imported in an asynchronously loaded lazy chunk, it throws at runtime.

### Link to Minimal Reproduction and step to  reproduce

https://github.com/Knagis/webpack-css-lazy-font-issue

### Expected Behavior

the asset resource is available in both CSS and JS chunks

### Actual Behavior

```
ERROR
Cannot find module './font.woff2'
    at __webpack_require__ (http://localhost:8080/main.js:216:21)
    at fn (http://localhost:8080/main.js:415:21)
    at eval (webpack:///./mod.js?:2:69)
    at ./mod.js (http://localhost:8080/mod_js.js:18:1)
    at __webpack_require__ (http://localhost:8080/main.js:221:32)
    at fn (http://localhost:8080/main.js:415:21)
```

### Environment

```shell
System:
    OS: Windows 11 10.0.26200
  Binaries:
    Node: 24.11.1
  Packages:
    html-webpack-plugin: ^5.6.6 => 5.6.6
    webpack: ^5.106.0 => 5.106.0
    webpack-cli: ^7.0.2 => 7.0.2
    webpack-dev-server: ^5.2.3 => 5.2.3
```

### Is this a regression?

None

### Last Working Version

_No response_

### Additional Context

also fails on webpack 5.104.1, possible earlier as well.

--- a/.changeset/fix-css-lazy-asset-resource.md
+++ b/.changeset/fix-css-lazy-asset-resource.md
@@ -1,0 +1,5 @@
+---
+"webpack": patch
+---
+
+Make asset modules available in JS context when referenced from both CSS and a lazily compiled JS chunk.

--- a/lib/NormalModule.js
+++ b/lib/NormalModule.js
@@ -1748,6 +1748,15 @@ class NormalModule extends Module {
 			/** @type {string} */
 			(buildInfo.hash)
 		);
+		// Clear cached source types and re-compute so that changes in incoming
+		// connections (e.g. asset module newly referenced from JS via lazy
+		// compilation) are reflected in the hash and trigger code generation
+		// cache invalidation.
+		// https://github.com/webpack/webpack/issues/20800
+		this._sourceTypes = undefined;
+		for (const type of this.getSourceTypes()) {
+			hash.update(type);
+		}
 		/** @type {Generator} */
 		(this.generator).updateHash(hash, {
 			module: this,

--- a/test/hotCases/css/lazy-compilation-asset/file.txt
+++ b/test/hotCases/css/lazy-compilation-asset/file.txt
@@ -1,0 +1,1 @@
+asset-content

--- a/test/hotCases/css/lazy-compilation-asset/index.js
+++ b/test/hotCases/css/lazy-compilation-asset/index.js
@@ -1,0 +1,13 @@
+import "./style.css";
+
+it("should make asset available in both CSS and lazy JS chunk", (done) => {
+	const promise = import("./mod.js");
+	NEXT_DEFERRED(
+		require("../../update")(done, true, () => {
+			promise.then((mod) => {
+				expect(mod.default).toContain("file.txt");
+				done();
+			}, done);
+		})
+	);
+});

--- a/test/hotCases/css/lazy-compilation-asset/mod.js
+++ b/test/hotCases/css/lazy-compilation-asset/mod.js
@@ -1,0 +1,2 @@
+import file from "./file.txt";
+export default file;

--- a/test/hotCases/css/lazy-compilation-asset/style.css
+++ b/test/hotCases/css/lazy-compilation-asset/style.css
@@ -1,0 +1,3 @@
+.font {
+	src: url("./file.txt");
+}

--- a/test/hotCases/css/lazy-compilation-asset/test.filter.js
+++ b/test/hotCases/css/lazy-compilation-asset/test.filter.js
@@ -1,0 +1,3 @@
+"use strict";
+
+module.exports = (config) => config.target === "web";

--- a/test/hotCases/css/lazy-compilation-asset/webpack.config.js
+++ b/test/hotCases/css/lazy-compilation-asset/webpack.config.js
@@ -1,0 +1,30 @@
+"use strict";
+
+/** @type {import("../../../../").Configuration} */
+module.exports = {
+	mode: "development",
+	devtool: false,
+	output: {
+		cssFilename: "[name].css",
+		cssChunkFilename: "[name].css",
+		assetModuleFilename: "assets/[name][ext]"
+	},
+	experiments: {
+		css: true,
+		lazyCompilation: {
+			entries: false,
+			imports: true
+		}
+	},
+	module: {
+		rules: [
+			{
+				test: /\.txt$/,
+				type: "asset/resource"
+			}
+		]
+	},
+	node: {
+		__dirname: false
+	}
+};

